kubernetes-exec-helper/src/Kubernetes/K8SChannel.hs

```diff
@@ -0,0 +1,113 @@
+{-# LANGUAGE OverloadedStrings #-}
+{-| 
+  Module : Kubernetes.K8SChannel
+  Description : This module maintains types to interact with STM channels and IO.
+-}
+module Kubernetes.K8SChannel where 
+
+import Control.Exception
+import Control.Exception.Safe
+import Control.Concurrent(ThreadId)
+import Control.Concurrent.Async(Async)
+import Control.Concurrent.STM
+import Data.Text
+import Kubernetes.KubeConfig
+import System.IO (stdin, stdout, stderr, Handle)
+
+-- | An interval in seconds for the timeout.
+type TimeoutInterval = Int
+
+-- | ChannelId encoding '[stdin, stdout, stderr]'.
+data ChannelId = StdIn | StdOut | StdErr | Error | Resize deriving (Eq, Ord, Enum)
+
+-- | An invalid channel.
+newtype InvalidChannel = InvalidChannel Text deriving (Show, Typeable)
+
+
+{- | 
+  CreateWSClient maintains all threads that are running,  
+  a writer channel to send messages to the server 
+  and a list of all "(ChannelId, TChan Text)" pairs.
+  Clients can wait on '[Async ThreadId]' and proceed to work with each channel.
+-}
+newtype CreateWSClient a = CreateWSClient {
```

Comment:
This is not really analogous to how System.Process works. In System.Process, the [CreateProcess](https://hackage.haskell.org/package/process-1.6.3.0/docs/System-Process.html#t:CreateProcess) structure contains information about the how to run the process: the executable, the arguments, the environment variables, etc, and the [createProcess](https://hackage.haskell.org/package/process-1.6.3.0/docs/System-Process.html#v:createProcess) function returns handles for manipulating stdin/stdout/stderr as well as the spawned process itself. The equivalent of CreateProcess here would be a structure that defines the command to execute and what container of what pod to execute that command in, and the equivalent of createProcess will return channels for reading the stdout/stderr of the command and sending data to stdin, as well as a handle for managing the websocket session itself.